src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,154 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import java.util.ArrayList;
+
+/**
+ * Interpret and process user inputs.
+ */
+public class Parser {
+    protected TaskList tasks;
+    protected Storage storage;
+    protected Ui ui;
+
+    /**
+     * Saves tasks, storage and ui class.
+     *
+     * @param tasks The TaskList class.
+     * @param storage The Storage class.
+     * @param ui The Ui class.
+     */
+    public Parser(TaskList tasks, Storage storage, Ui ui) {
+        this.tasks = tasks;
+        this.storage = storage;
+        this.ui = ui;
+    }
+
+    /**
+     * Processes the user input.
+     *
+     * @param userInput The commands issued to duke by the user.
+     * @return Returns false when user wants to terminate the program.
+     */
+    public boolean parse(String userInput) {
+        if (userInput.equals("bye")) {
+            return false;
+        }
+        String data;
+
+        String[] inputs = userInput.trim().split(" ", 2);
+        switch (inputs[0]) {
+        case "list":
+            tasks.list();
+            break;
+        case "done":
+        case "delete":
+            try {
+                int taskNumber = Integer.parseInt(inputs[1].trim()) - 1;
+                if (inputs[0].equals("done")) {
+                    try {
+                        tasks.get(taskNumber).markAsDone();
+                    } catch (DukeException e) {
+                        ui.exceptionMessage(e);
+                        break;
+                    }
+                    ui.userMessage("Nice! I've marked this task as done:\n    " + tasks.get(taskNumber));
+                } else {
+                    Task deletedTask = tasks.remove(taskNumber);
+                    ui.userMessage("Noted. I've removed this task: \n    " + deletedTask
+                            + "\nNow you have " + tasks.size() + " tasks in the list.");
+                }
+                storage.writeToFile(inputs[0], taskNumber);
+            } catch (Exception e) {
+                ui.exceptionMessage(new DukeException("☹ OOPS!!! Please provide a task number within range."));
+            }
+            break;
```

Comment:
Maybe it's just me, but I do not know the purpose of the "done" case as it's empty, and how your "delete" case does what the "done" command is supposed to do too.